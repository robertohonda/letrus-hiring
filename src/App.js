import React, { Component } from 'react'
import './App.css'

export const sortMistakes = (mistakes) => {
  const mistakesCopy = [...mistakes]
  return mistakesCopy.sort(
    (mistake, nextMistake) =>
      mistake.paragraph - nextMistake.paragraph ||
      mistake.start - nextMistake.start,
  )
}

export const makeTextWithMistakes = (text, mistakes) => {
  const texts = text.split('\n')
  const sortedMistakes = sortMistakes(mistakes)
  const result = texts.map((text) => [text])
  const startIndexes = new Array(texts.length).fill(0)

  sortedMistakes.forEach((mistake, mistakeIndex) => {
    const { start, end, paragraph } = mistake
    const text = texts[paragraph]
    const seletectedParagraph = result[paragraph]
    const selectedStart = startIndexes[paragraph]
    const newTexts = []

    seletectedParagraph.pop()

    if (start > selectedStart) newTexts.push(text.slice(selectedStart, start))
    newTexts.push(<em key={mistakeIndex}>{text.slice(start, end)}</em>)
    if (end < text.length) newTexts.push(text.slice(end))

    seletectedParagraph.push(...newTexts)
    startIndexes[paragraph] = end
  })

  return result
}

class App extends Component {
  renderParagraph = (paragraph, i) => {
    return <p key={i}>{paragraph}</p>
  }

  render() {
    const text = 'Meu texto está erado\nSegundo palagrafo'
    const mistakes = [
      {
        start: 15,
        end: 20,
        paragraph: 0,
      },
      {
        start: 8,
        end: 17,
        paragraph: 1,
      },
    ]

    // ========================================================================
    // Sua tarefa é implementar a função que converte as variáveis `text` e
    // `mistakes` na variável `textWithMistakes`, para textos e erros
    // arbitrários. Você deve fazer isso na função
    // `makeTextWithMistakes(text, mistakes)` usando outras funções de ajuda se
    // achar necessário
    // ========================================================================

    // const textWithMistakes = [
    //   ['Meu texto está ', <em>erado</em>],
    //   ['Segundo ', <em>palagrafo</em>],
    // ]

    const textWithMistakes = makeTextWithMistakes(text, mistakes)

    return (
      <div className="App">{textWithMistakes.map(this.renderParagraph)}</div>
    )
  }
}

export default App
