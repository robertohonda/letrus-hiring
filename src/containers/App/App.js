import React from 'react'
import theme from '../../theme'
import { ThemeProvider } from '@material-ui/styles'
import CssBaseline from '@material-ui/core/CssBaseline'
import AppBar from '../../components/AppBar'
import Typography from '@material-ui/core/Typography'

import { makeStyles } from '@material-ui/core/styles'

export const sortMistakes = (mistakes) => {
  const mistakesCopy = [...mistakes]
  return mistakesCopy.sort(
    (mistake, nextMistake) =>
      mistake.paragraph - nextMistake.paragraph ||
      mistake.start - nextMistake.start,
  )
}

export const makeTextWithMistakes = (text, mistakes) => {
  const texts = text.split('\n')
  const sortedMistakes = sortMistakes(mistakes)
  const result = texts.map((text) => [text])
  const startIndexes = new Array(texts.length).fill(0)

  sortedMistakes.forEach((mistake, mistakeIndex) => {
    const { start, end, paragraph } = mistake
    const text = texts[paragraph]
    const seletectedParagraph = result[paragraph]
    const selectedStart = startIndexes[paragraph]
    const newTexts = []

    seletectedParagraph.pop()

    if (start > selectedStart) newTexts.push(text.slice(selectedStart, start))
    newTexts.push(
      <Typography key={mistakeIndex} component="em" color="error">
        {text.slice(start, end)}
      </Typography>,
    )
    if (end < text.length) newTexts.push(text.slice(end))

    seletectedParagraph.push(...newTexts)
    startIndexes[paragraph] = end
  })

  return result
}

const useStyles = makeStyles((theme) => ({
  main: {
    padding: theme.spacing(2),
  },
  toolbar: theme.mixins.toolbar,
}))

const App = (props) => {
  const classes = useStyles(props)
  const text = 'Meu texto estÃ¡ erado\nSegundo palagrafo'
  const mistakes = [
    {
      start: 15,
      end: 20,
      paragraph: 0,
    },
    {
      start: 8,
      end: 17,
      paragraph: 1,
    },
  ]
  const textWithMistakes = makeTextWithMistakes(text, mistakes)

  const renderParagraph = (paragraph, i) => {
    return <p key={i}>{paragraph}</p>
  }

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <AppBar logoPath="logo.svg" />
      <main className={classes.main}>
        <div className={classes.toolbar} />
        <div>{textWithMistakes.map(renderParagraph)}</div>
      </main>
    </ThemeProvider>
  )
}

export default App
