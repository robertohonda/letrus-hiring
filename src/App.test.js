import React from 'react'
import { shallow } from 'enzyme'
import App, { makeTextWithMistakes } from './App'

describe('App', () => {
  it('Should render correctly', () => {
    const component = shallow(<App />)

    expect(component).toMatchSnapshot()
  })
})

describe('makeTextWithMistakes', () => {
  const mocks = [
    {
      text: 'Meu texto está erado\nSegundo palagrafo',
      mistakes: [
        {
          start: 15,
          end: 20,
          paragraph: 0,
        },
        {
          start: 8,
          end: 17,
          paragraph: 1,
        },
      ],
      output: [
        ['Meu texto está ', <em key={0}>erado</em>],
        ['Segundo ', <em key={1}>palagrafo</em>],
      ],
    },
    {
      text:
        'Todo aluno pode aprender a ler e a escrever plenamente, sem exceções.',
      mistakes: [],
      output: [
        [
          'Todo aluno pode aprender a ler e a escrever plenamente, sem exceções.',
        ],
      ],
    },
    {
      text:
        'O letlamento impazcta na autoeztima e confianza do aluno e tem conzequências em todo o cicro de aprendiçagem escolar, em todas as diziplinas.\nA esclita é um moztor ezencial para ampriar nossas perzepções de mundo e, assim, ampriar nossas capazidades de ação e escolhas.\nAcleditamos que a capazidade de se explessar é uma competênzia fundamental ao desenvorvimento cognitivo, humano e social de todo individo.',
      mistakes: [
        {
          start: 2,
          end: 21,
          paragraph: 0,
        },
        {
          start: 25,
          end: 35,
          paragraph: 0,
        },
        {
          start: 38,
          end: 47,
          paragraph: 0,
        },
        {
          start: 63,
          end: 75,
          paragraph: 0,
        },
        {
          start: 87,
          end: 92,
          paragraph: 0,
        },
      ],
      output: [],
    },
  ]

  it('Should return text with mistakes correctly', () => {
    mocks.forEach((mock) => {
      const { text, mistakes, output } = mock
      const result = makeTextWithMistakes(text, mistakes)
      expect(result).toEqual(output)
    })
  })
})
