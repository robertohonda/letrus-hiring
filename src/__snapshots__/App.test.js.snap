// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render correctly 1`] = `
<div
  className="App"
>
  <p
    key="0"
  >
    Meu texto est√° 
    <em
      key="0"
    >
      erado
    </em>
  </p>
  <p
    key="1"
  >
    Segundo 
    <em
      key="1"
    >
      palagrafo
    </em>
  </p>
</div>
`;
